File: syslog.pl
/*  Part of SWI-Prolog

    Author:        Jan Wielemaker
    E-mail:        J.Wielemaker@cs.vu.nl
    WWW:           http://www.swi-prolog.org
    Copyright (C): 2013, VU University Amsterdam

    This program is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public License
    as published by the Free Software Foundation; either version 2
    of the License, or (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public
    License along with this library; if not, write to the Free Software
    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

    As a special exception, if you link this library with other files,
    compiled with a Free Software compiler, to produce an executable, this
    library does not by itself cause the resulting executable to be covered
    by the GNU General Public License. This exception does not however
    invalidate any other reasons why the executable file might be covered by
    the GNU General Public License.
*/

:- module(syslog,
	  [ openlog/3,			% +Ident, +Options, +Facility
	    syslog/2,			% +Priority, +Message
	    syslog/3,			% +Priority, +Format, +Args
	    closelog/0
	  ]).

/** <module> Unix syslog interface

This library provides an interface to   the  Unix syslog() facility. The
interface is an almost direct translation of  the POSIX syslog API, with
two additions:

  - syslog/3 exploits format/3 to format syslog messages
  - The library integrates into library(debug) using
    prolog:debug_print_hook/3, where debug _topics_ are mapped to
    syslog _priorities_ and remaining debug _topics_ are mapped
    to the syslog _priority_ =debug=.

Note that this interface  makes  no   attempt  to  abstract over logging
facilities of operating systems. We expect   that such abstractions will
be implemented at the Prolog  level   using  multiple  integrations into
library(debug).

@see	detach_IO/1 to detach normal I/O of the process and remove it
	from the process group.
@see	fork/1 to create a daemon process.
@see	library(uid) to manage user identifiers (e.g., drop root
	privileges).
*/

:- use_foreign_library(foreign(syslog)).

:- dynamic syslog/1.

%%	openlog(+Ident:atom, +Options:list(atom), +Facility:atom) is det.
%
%	Open system log. This predicate provides a direct interface into
%	the openlog() library call. If the   library call is successful,
%	it runs at_halt(closelog) to ensure closing   the  system log on
%	clean exit.
%
%	@param	Ident prepended to every message, and is typically set
%		to the program name.
%	@param	Options is a list of options.  Values are corresponding
%		C options, after removing =LOG_= and translation to
%		lower case: =cons=, =ndelay=, =nowait=, =odelay=,
%		=perror=, =pid=.
%	@param	Facility is one of =auth=, =authpriv=, =cron=, =daemon=,
%		=ftp=, =kern=, =local0= ... =local7=, =lpr=, =mail=,
%		=news=, =syslog=, =user= or =uucp=.

openlog(Ident, Options, Facility) :-
	'$openlog'(Ident, Options, Facility),
	asserta(syslog(Ident)),
	at_halt(closelog).

%%	syslog(+Priority, +Message) is det.
%
%	Send a message to the system  log. Note that syslog/2 implicitly
%	opens a connection to the system log   if  such a connection has
%	not been opened explicitly using openlog/3.
%
%	@param	Priority is one of =emerg=, =alert=, =crit=, =err=,
%		=warning=, =notice=, =info= or =debug=.

%%	syslog(+Priority, +Format, +Args) is det.
%
%	Send a formatted message to the system  log if system logging is
%	opened using openlog/3. This predicate   combined  format/3 with
%	syslog/2. If there is no open  syslog connection, syslog/3 calls
%	print_message/2.

syslog(Priority, Format, Args) :-
	syslog(_), !,
	format(string(Msg), Format, Args),
	syslog(Priority, Msg).
syslog(Priority, Format, Args) :-
	syslog_priority(Priority, Kind),
	print_message(Kind, format(Format, Args)).

%%	closelog is det.
%
%	Close the system log.

closelog :-
	retractall(syslog(_)),
	'$closelog'.


		 /*******************************
		 *     DEBUG INTEGRATION	*
		 *******************************/

:- multifile
	prolog:debug_print_hook/3.

%%	prolog:debug_print_hook(+Topic, +Format, +Args) is semidet.
%
%	Integration of debug/3 with the syslog   facility.  If syslog is
%	enabled, debug/3 is re-routed to use   the syslog facilities. If
%	the _topic_ of the  debug  message   matches  one  of  the sylog
%	_priority_ values (see syslog/2), the message   is sent with the
%	corresponding syslog priority. Otherwise  it   it  sent with the
%	=debug= priority.

prolog:debug_print_hook(Topic, Format, Args) :-
	syslog(_),
	debug_priority(Topic, Priority),
	syslog(Priority, Format, Args).

debug_priority(Topic, Priority) :-
	(   syslog_priority(Topic, _Kind)
	->  Priority = Topic
	;   Priority = debug
	).

syslog_priority(emerg,	 error).
syslog_priority(alert,	 warning).
syslog_priority(crit,	 error).
syslog_priority(err,	 error).
syslog_priority(warning, warning).
syslog_priority(notice,	 informational).
syslog_priority(info,	 informational).
syslog_priority(debug,	 debug).


		 /*******************************
		 *	MESSAGE INTEGRATION	*
		 *******************************/

user:message_hook(Term, Kind, _) :-
	syslog(_),
	kind_syslog_priority(Kind, Level),
	message_to_string(Term, Message),
	atomic_list_concat(Lines, '\n', Message),
	forall(member(Line, Lines),
	       syslog(Level, Line)),
	fail.

kind_syslog_priority(error,	    err).
kind_syslog_priority(warning,	    warning).
kind_syslog_priority(informational, info).

File: syslog.c
/*  Part of SWI-Prolog

    Author:        Jan Wielemaker
    E-mail:        J.Wielemaker@vu.nl
    WWW:           http://www.swi-prolog.org
    Copyright (c)  2013-2015, VU University Amsterdam
    All rights reserved.

    Redistribution and use in source and binary forms, with or without
    modification, are permitted provided that the following conditions
    are met:

    1. Redistributions of source code must retain the above copyright
       notice, this list of conditions and the following disclaimer.

    2. Redistributions in binary form must reproduce the above copyright
       notice, this list of conditions and the following disclaimer in
       the documentation and/or other materials provided with the
       distribution.

    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
    FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
    COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
    INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
    BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
    POSSIBILITY OF SUCH DAMAGE.
*/

#include <config.h>
#include <SWI-Prolog.h>
#include <syslog.h>
#include <string.h>

#define streq(s,q) (strcmp(s,q) == 0)

static int
get_option(term_t t, int *opt)
{ term_t tail = PL_copy_term_ref(t);
  term_t head = PL_new_term_ref();
  char *s;
  int option = 0;

  while( PL_get_list_ex(tail, head, tail) )
  { if ( PL_get_chars(head, &s, CVT_ATOM|CVT_EXCEPTION) )
    { if      ( streq(s, "cons"  ) ) option |= LOG_CONS;
      else if ( streq(s, "ndelay") ) option |= LOG_NDELAY;
      else if ( streq(s, "nowait") ) option |= LOG_NOWAIT;
      else if ( streq(s, "odelay") ) option |= LOG_ODELAY;
#ifdef LOG_PERROR
      else if ( streq(s, "perror") ) option |= LOG_PERROR;
#endif
      else if ( streq(s, "pid") )    option |= LOG_PID;
      else return PL_domain_error("syslog_option", head);
    } else
      return FALSE;
  }

  if ( PL_get_nil_ex(tail) )
  { *opt = option;
    return TRUE;
  }

  return FALSE;
}

static int
get_facility(term_t t, int *fac)
{ char *s;
  int facility;

  if ( PL_get_chars(t, &s, CVT_ATOM|CVT_EXCEPTION) )
  { if      ( streq(s, "auth"  ) )   facility = LOG_AUTH;
#ifdef LOG_AUTHPRIV
    else if ( streq(s, "authpriv") ) facility = LOG_AUTHPRIV;
#endif
    else if ( streq(s, "cron") )     facility = LOG_CRON;
    else if ( streq(s, "daemon") )   facility = LOG_DAEMON;
#ifdef LOG_FTP
    else if ( streq(s, "ftp") )      facility = LOG_FTP;
#endif
    else if ( streq(s, "kern") )     facility = LOG_KERN;
    else if ( streq(s, "local0") )   facility = LOG_LOCAL0;
    else if ( streq(s, "local1") )   facility = LOG_LOCAL1;
    else if ( streq(s, "local2") )   facility = LOG_LOCAL2;
    else if ( streq(s, "local3") )   facility = LOG_LOCAL3;
    else if ( streq(s, "local4") )   facility = LOG_LOCAL4;
    else if ( streq(s, "local5") )   facility = LOG_LOCAL5;
    else if ( streq(s, "local6") )   facility = LOG_LOCAL6;
    else if ( streq(s, "local7") )   facility = LOG_LOCAL7;
    else if ( streq(s, "lpr") )      facility = LOG_LPR;
    else if ( streq(s, "mail") )     facility = LOG_MAIL;
    else if ( streq(s, "news") )     facility = LOG_NEWS;
    else if ( streq(s, "syslog") )   facility = LOG_SYSLOG;
    else if ( streq(s, "user") )     facility = LOG_USER;
    else if ( streq(s, "uucp") )     facility = LOG_UUCP;
    else return PL_domain_error("syslog_facility", t);
  } else
    return FALSE;

  *fac = facility;
  return TRUE;
}

static int
get_priority(term_t t, int *pri)
{ char *s;
  int priority;

  if ( PL_get_chars(t, &s, CVT_ATOM|CVT_EXCEPTION) )
  { if      ( streq(s, "emerg"  ) ) priority =	LOG_EMERG;
    else if ( streq(s, "alert") )   priority = LOG_ALERT;
    else if ( streq(s, "crit") )    priority = LOG_CRIT;
    else if ( streq(s, "err") )     priority = LOG_ERR;
    else if ( streq(s, "warning") ) priority = LOG_WARNING;
    else if ( streq(s, "notice") )  priority = LOG_NOTICE;
    else if ( streq(s, "info") )    priority = LOG_INFO;
    else if ( streq(s, "debug") )   priority = LOG_DEBUG;
    else
    { PL_domain_error("syslog_priority", t);
      return FALSE;
    }
  } else
    return FALSE;

  *pri = priority;
  return TRUE;
}



static foreign_t
pl_openlog(term_t Ident, term_t Option, term_t Facility)
{ char *ident;
  int option = 0;
  int facility = 0;

  if ( PL_get_chars(Ident, &ident, CVT_ATOM|REP_MB|CVT_EXCEPTION) &&
       get_option(Option, &option) &&
       get_facility(Facility, &facility) )
  { openlog(strdup(ident), option, facility);
    return TRUE;
  }

  return FALSE;
}


static foreign_t
pl_syslog(term_t Priority, term_t Message)
{ int p;
  char *msg;

  if ( get_priority(Priority, &p) &&
       PL_get_chars(Message, &msg,
		    CVT_ALL|CVT_VARIABLE|CVT_WRITE|REP_MB|CVT_EXCEPTION) )
  { syslog(p, "%s", msg);

    return TRUE;
  }

  return FALSE;
}


static foreign_t
pl_closelog(void)
{ closelog();

  return TRUE;
}


install_t
install_syslog(void)
{ PL_register_foreign("$openlog",  3, pl_openlog,  0);
  PL_register_foreign("syslog",    2, pl_syslog,   0);
  PL_register_foreign("$closelog", 0, pl_closelog, 0);
}
